### PR TITLE
feat: support regex automod

### PR DESCRIPTION
**Pull request by onerandomusername**
## Summary

implements https://github.com/discord/discord-api-docs/pull/5504

This has been tested in a private server.

## Checklist

- [ ] If code changes were made, then they have been tested
    - [x] I have updated the documentation to reflect the changes
    - [x] I have formatted the code properly by running `task lint`
    - [x] I have type-checked the code by running `task pyright`
- [ ] This PR fixes an issue
- [x] This PR adds something new (e.g. new method or parameters)
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: disnake/automod.py [hunk 1]
```diff
@@ -244,6 +256,14 @@ class AutoModTriggerMetadata:
     def __init__(self, *, keyword_filter: Sequence[str]):
         ...
 
+    @overload
+    def __init__(self, *, regex_patterns: Sequence[str]):
+        ...
+
+    @overload
+    def __init__(self, *, keyword_filter: Sequence[str], regex_patterns: Sequence[str]):
+        ...
```
Comment (shiftinv) [inline disnake/automod.py:265]:
```suggestion
    def __init__(self, *, keyword_filter: Sequence[str], regex_patterns: Sequence[str] = ...):
        ...

    @overload
    def __init__(self, *, keyword_filter: Sequence[str] = ..., regex_patterns: Sequence[str]):
        ...
```

File: changelog/794.feature.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+Support regex within automod. Adds :attr:`AutoModTriggerMetadata.regex_patterns`
```
Comment (shiftinv) [inline changelog/794.feature.rst:1]:
```suggestion
Support regex within automod using :attr:`AutoModTriggerMetadata.regex_patterns`.
```

**Review by shiftinv — CHANGES_REQUESTED** [review]

**Comment by shiftinv** [comment]
The documentation for `AutoModActionExecution.matched_keyword` should mention regular expressions as well, looks like that field is being reused:
```json
{
    "t": "AUTO_MODERATION_ACTION_EXECUTION",
    "s": 12,
    "op": 0,
    "d": {
        "user_id": "",
        "rule_trigger_type": 1,
        "rule_id": "984891789930557500",
        "matched_keyword": "literally 1\\s*9\\s*8\\s*4",
        "matched_content": "literally 1 9 8 4",
        "content": "wow this is literally 1 9 8 4 smh",
        ...
    }
}
```

**Comment by onerandomusername** [comment]
@shiftinv I added a mention of matched_keyword possibly containing regex in https://github.com/DisnakeDev/disnake/pull/794/commits/5805f24e91be22daa4d22174feb18348a9bff3e6

File: disnake/automod.py [hunk 3]
```diff
@@ -614,6 +639,11 @@ class AutoModActionExecution:
 
     matched_keyword: Optional[:class:`str`]
         The keyword that matched.
+
+        .. note::
+
+            If this rule matched because of a regex, this will contain the matching regex.
```
Comment (shiftinv) [inline disnake/automod.py:645]:
```suggestion
        The keyword or regex that matched.
```

**Review by shiftinv — APPROVED** [review]

File: disnake/automod.py [hunk 4]
```diff
@@ -223,6 +223,17 @@ class AutoModTriggerMetadata:
         See :ddocs:`api docs <resources/auto-moderation#auto-moderation-rule-object-keyword-matching-strategies>`
         for details about how keyword matching works.
 
+    regex_patterns: Optional[Sequence[:class:`str`]]
+        The list of regular expressions to check for. Used with :attr:`AutoModTriggerType.keyword`.
+
+        A maximum of 10 regexes can be added.
```
Comment (shiftinv) [inline disnake/automod.py:229]:
```suggestion
        A maximum of 10 regexes can be added, each with up to 75 characters.
```

File: disnake/automod.py [hunk 5]
```diff
@@ -223,6 +223,17 @@ class AutoModTriggerMetadata:
         See :ddocs:`api docs <resources/auto-moderation#auto-moderation-rule-object-keyword-matching-strategies>`
         for details about how keyword matching works.
```
Comment (shiftinv) [inline disnake/automod.py:224]:
```suggestion
        for details about how keyword matching works.
        Each keyword must be 30 characters or less.
```

**Pull request merged**